### ISSUE TITLE
[explicit-module-boundary-types] Complains about typing in nested interfaces

### ISSUE THREAD

**Issue by essenmitsosse** [typescript-eslint/typescript-eslint#2214]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-function-return-type": [2, {"allowExpressions": true}],
    "@typescript-eslint/explicit-module-boundary-types": [2]
  }
}
```

```TS
export const foo: {
    interfaceOuter: {
        interfaceInner: (input: number) => number;
    };
} = {
    interfaceOuter: {
        interfaceInner: (input) => (input),
    },
};
```

**Expected Result**
Shouldn't give an error for any function where the return type is in any way explicitly typed.

**Actual Result**
Throws errors if function type is declared in a nested interface

```
Missing return type on function   @typescript-eslint/explicit-module-boundary-types
Argument 'input' should be typed  @typescript-eslint/explicit-module-boundary-types
```

**Additional Info**

Reproduction repo: https://github.com/essenmitsosse/typescript-eslint-explicit-module-boundary-types

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.2.0` |
| `@typescript-eslint/parser`        | `3.2.0` |
| `TypeScript`                       | `3.9.3` |
| `ESLint`                           | `7.1.0` |
| `node`                             | `13.14.0` |
| `npm`                              | `6.14.4` |

**Comment by bradzacher** [comment]
This is the same issue as in #2214
